allennlp/models/encoder_decoders/wikitables_semantic_parser.py

```diff
@@ -275,93 +389,109 @@ def __init__(self,
-                # keep track of when the top action wasn't allowed.  NOTE: without batching, when
-                # there are ~80 possible actions in each state, this gives a ~2x slowdown.  So I'm
-                # keeping it for now.  I'll re-evaluate once batching is implemented.
-                continue
-            new_action_history = state.action_history + [action]
-            # TODO(matt): we might need to normalize this by length eventually.  If the training
-            # algorithm allows for comparing states at different timesteps, it might matter.
-            new_score = state.score + sorted_log_probs[0, i]
-            yield WikiTablesDecoderState(new_action_history,
-                                         new_score,
-                                         state.encoder_outputs,
-                                         state.encoder_output_mask,
-                                         new_hidden_state,
-                                         state.end_index)
+        for group_index in range(group_size):
+            for action_index in range(num_actions):
+                action = sorted_actions[group_index][action_index]
+                if valid_actions is not None and action not in valid_actions[group_index]:
+                    # If we have a constrained decoder, we need to skip any actions that are not valid
+                    # in the current state.
+                    continue
+                if allowed_actions is not None and action not in allowed_actions[group_index]:
+                    # This happens when our _decoder trainer_ wants us to only evaluate certain
+                    # actions, likely because they are the gold actions in this state.  We just skip
+                    # emitting any state that isn't allowed by the trainer, because constructing the
+                    # new state can be expensive.
+                    # TODO(mattg): is this really all that expensive?  If it isn't, it would simplify
+                    # the API to remove the `allowed_actions` argument, and it would make computing
+                    # metrics easier if this skipping logic happened in the trainer - the trainer could
+                    # keep track of when the top action wasn't allowed.  NOTE: without batching, when
+                    # there are ~80 possible actions in each state, this gives a ~2x slowdown.  So I'm
+                    # keeping it for now.  I'll re-evaluate once batching is implemented.
+                    continue
+                # We'll yield a bunch of states here that all have a `group_size` of 1, so that the
+                # learning algorithm can decide how many of these it wants to keep, and it can just
+                # regroup them later, as that's a really easy operation.
+                new_action_history = state.action_history[group_index] + [action]
+                # TODO(matt): we might need to normalize this by length eventually.  If the training
```

Comment:
I don't think this is too problematic - the `WikiTablesDecoderState` contains everything required to normalise with respect to its action sequence, so how the learning algorithm/search proceedure decides to expand nodes doesn't require any action here, I don't think. Seems like you can just remove this `TODO`.

Reply:
Done.